### ISSUE TITLE
Placeholder for SelectOneMenu

### ISSUE THREAD

**Issue by RueKow** [primefaces/primefaces#2191]
## 1) Environment
- PrimeFaces version: 6.1 RC2
- Does it work on the newest released PrimeFaces version? No
- Does it work on the newest sources in GitHub? No
- Application server + version: Tomcat 8
- Affected browsers: all

## 2) Expected behavior
A stylable hint should be displayed if selected item is empty or null.

## 3) Actual behavior
There is no placeholder attribute available.

## 4) Steps to reproduce
There is no placeholder attribute available.
See forum post: https://forum.primefaces.org/viewtopic.php?f=3&t=50027

## 5) Sample XHTML
```
<p:selectOneMenu id="somPunktrunde001" value="#{publicController.day}"
   style="width:200px" placeholder="Bitte wählen ...">
   <f:selectItem itemLabel=""           itemValue="" />
   <f:selectItem itemLabel="Montag"     itemValue="1" />
   <f:selectItem itemLabel="Dienstag"   itemValue="2" />
   <f:selectItem itemLabel="Mittwoch"   itemValue="3" />
   <f:selectItem itemLabel="Donnerstag" itemValue="4" />
   <f:selectItem itemLabel="Freitag"    itemValue="5" />
   <f:selectItem itemLabel="Samstag"    itemValue="6" />
   <f:selectItem itemLabel="Sonntag"    itemValue="7" />
</p:selectOneMenu>
```

## 6) Sample bean
```
@ManagedBean
@SessionScoped
public class PublicController implements Serializable
{
	private String day;

	@PostConstruct
	public void init()
	{
		day = "";
	}

	public String getDay()
	{
		return day;
	}

	public void setDay(String day)
	{
		this.day = day;
	}

}
```

## 7) Proposal to realize it
Modified SelectOneMenu class (new PropertyKey "placeholder"):
```
@ResourceDependencies(
{ @ResourceDependency(library = "primefaces", name = "components.css"),
		@ResourceDependency(library = "primefaces", name = "jquery/jquery.js"),
		@ResourceDependency(library = "primefaces", name = "jquery/jquery-plugins.js"),
		@ResourceDependency(library = "primefaces", name = "core.js"),
		@ResourceDependency(library = "primefaces", name = "components.js") })
public class SelectOneMenu extends HtmlSelectOneMenu
		implements org.primefaces.component.api.Widget, org.primefaces.component.api.InputHolder
{

	public enum PropertyKeys
	{

		widgetVar, effect, effectSpeed, panelStyle, panelStyleClass, var, height, editable, filter, 
		filterMatchMode, filterFunction, caseSensitive, maxlength, appendTo, title, syncTooltip, 
		labelTemplate, autoWidth, placeholder;

		...

	}

	public java.lang.String getPlaceholder()
	{
		return (java.lang.String) getStateHelper().eval(PropertyKeys.placeholder, null);
	}

	public void setPlaceholder(java.lang.String _placeholder)
	{
		getStateHelper().put(PropertyKeys.placeholder, _placeholder);
	}

	...

}
```
Modified SelectOneMenuRenderer class (method "encodeLabel" extended by "placeholder"):
```
public class SelectOneMenuRenderer extends SelectOneRenderer
{

	...

	protected void encodeLabel(FacesContext context, SelectOneMenu menu, List<SelectItem> selectItems)
			throws IOException
	{
		ResponseWriter writer = context.getResponseWriter();
		String valueToRender = ComponentUtils.getValueToRender(context, menu);

		if (menu.isEditable())
		{
			writer.startElement("input", null);
			writer.writeAttribute("type", "text", null);
			writer.writeAttribute("name", menu.getClientId(context) + "_editableInput", null);
			writer.writeAttribute("class", SelectOneMenu.LABEL_CLASS, null);

			if (menu.getTabindex() != null)
			{
				writer.writeAttribute("tabindex", menu.getTabindex(), null);
			}

			if (menu.isDisabled())
			{
				writer.writeAttribute("disabled", "disabled", null);
			}

			if (valueToRender != null)
			{
				writer.writeAttribute("value", valueToRender, null);
			}

			if (menu.getMaxlength() != Integer.MAX_VALUE)
			{
				writer.writeAttribute("maxlength", menu.getMaxlength(), null);
			}

			if (menu.getPlaceholder() != null)
			{
				writer.writeAttribute("placeholder", menu.getPlaceholder(), null);
			}

			writer.endElement("input");
		}
		else
		{
			writer.startElement("label", null);
			writer.writeAttribute("id", menu.getClientId(context) + "_label", null);
			writer.writeAttribute("class", SelectOneMenu.LABEL_CLASS, null);
			if (menu.getPlaceholder() != null)
			{
				writer.writeAttribute("placeholder", menu.getPlaceholder(), null);
			}
			writer.write("&nbsp;");
			writer.endElement("label");
		}
	}

	...

}
```
Modified PrimeFaces Widget Javascript (extended setLabel function):
```
/**
 * PrimeFaces SelectOneMenu Widget
 */
if (PrimeFaces.widget.SelectOneMenu)
{
    PrimeFaces.widget.SelectOneMenu = PrimeFaces.widget.SelectOneMenu.extend(
    {
        init: function(cfg) 
        {
            this._super(cfg);
        },

        setLabel: function(value)
        {
	        var displayedLabel = this.getLabelToDisplay(value);
	        if (this.cfg.editable) {
	            if (value === '&nbsp;')
	                this.label.val('');
	            else
	                this.label.val(displayedLabel);
	        }
	        else {
	            if (this.label.attr('placeholder') == null)
	            	var labelText = '&nbsp;';
	            else
	            	var labelText = this.label.attr('placeholder');
	            if (value === '&nbsp;') {
	            	this.label.addClass('ui-watermark');
	                this.label.html(labelText);
	            }
	            else {
	            	this.label.removeClass('ui-watermark');
	                this.label.text(displayedLabel);
	        	}
	        }
        }
    });
};
```

**Comment by Rapster** [comment]
Please provide a PR instead

**Comment by RueKow** [comment]
I never used git in my projects, always subversion. So I don't know how to provide a pull request.

**Comment by Rapster** [comment]
Time to start then, it's very easy really https://help.github.com/articles/creating-a-pull-request/

**Comment by RueKow** [comment]
Ok, I will try. Thanks for the link.

**Comment by RueKow** [comment]
Not so easy as mentioned. I studied GitHub Flow for better understanding and I installed GitHub Desktop application on my Mac. Then I tried to follow the steps of GitHub Flow:

1. Create a feature branch: ok
2. Add commits: 
3. Open a Pull Request: 

When trying to add commits I could not found "SelectOneMenu.java" and "component.js" in the PrimeFaces repository. Only "SelectOneMenuRenderer.java" is avilable.

What !'m doing wrong?

**Comment by Rapster** [comment]
I'm not familiar with Github Desktop, I'm using the good old TurtoistGit :) I suggest you to ask those questions to Github Desktop forum

**Comment by RueKow** [comment]
The problem is not GitHub Desktop.
I cannot find "SelectOneMenu.java" and "component.js" in the repository.

**Comment by Rapster** [comment]
You have to generates sources, doing `mvn clean install` should do ;)

**Comment by tandraschko** [comment]
Yep, component classes are generated (see src\main\java-templates) and components.js is also generated (see \src\main\resources\META-INF\resources\primefaces for the single files)

**Comment by melloware** [comment]
@RueKow I will take this one and submit a PR.

**Comment by melloware** [comment]
PR submitted I made one slight modification to your JS code.

**Comment by tandraschko** [comment]
AFAICS you added new placeholder attribute to the label, too. Why?

**Comment by tandraschko** [comment]
(i'm asking because i think that "placeholder" is not a valid attribute for a label element)

**Comment by melloware** [comment]
You are right I just updated the PR and removed it.  It was in @RueKow's code so I assumed it had some value but looking at the HTML5 docs it can't be used on a label!   Good catch!

**Comment by tandraschko** [comment]
I still think there is something wrong.
It tries to get the placeholder from the label: var labelText = this.label.attr('placeholder');

Also i'm not sure why a "ui-watermark" is required. Also, AFAIR there are some cases when the there is a label used instead input.

**Comment by melloware** [comment]
You are right I guess @Ruekow idea was if the label was being displayed it would show the placeholder value instead of blank.  But that really doesn't make sense the only thing this could should do is add the HTML placeholder to the Input.

**Comment by melloware** [comment]
That is what it is doing.  When the label is used if this is a read-only or disabled SelectOneMenu he must want to display the placeholder value.  That makes sense to me.

**Comment by tandraschko** [comment]
thats ok but as i said, placeholder is not allowed on label. So it should be a data attribute instead.

**Comment by melloware** [comment]
Good call i updated to "data-placeholder" for lack of a better data- name?

**Comment by tandraschko** [comment]
Yep, but please use jQuery's data method instead attr.
Also, please check that this code is only called when the label AND not the input is displayed.

**Comment by melloware** [comment]
OK i tested every permutation with my latest changes.

1. Normal SelectOne with <f:selectItem itemLabel="Select One" itemValue="" /> displays "Select One" without the watermark entirely.

2. editable="false" with placeholder="Non Editable" and <f:selectItem itemLabel="" itemValue="" /> show the label styled to look like a watermark.

3. editable="true" with placeholder="Editable" and <f:selectItem itemLabel="" itemValue="" /> show the HTML5 placeholder. 


![image](https://cloud.githubusercontent.com/assets/4399574/24711783/f3dc442c-19ee-11e7-8c65-35180ae99757.png)

**Comment by tandraschko** [comment]
merged it for 6.1 as it's a very small feature. Thanks.
